vault/token_store.go

```diff
@@ -2441,6 +2451,11 @@ func (ts *TokenStore) handleCreateCommon(ctx context.Context, req *logical.Reque
 		}
 	}
 
+	if role != nil && role.AllowGlobPolicies {
```

Comment:
this is the only bit in the PR ( outside of tests, which i haven't looked at closely yet ) that I'm not sure about, could you write down your reasoning for removing these policies ?

Reply:
If a token is created from a role and no policies are specified during the create the list of policies on the token defaults to the list of allowed_policies as defined on the role. This logic occurs here https://github.com/hashicorp/vault/blob/d6fea845bfb6b5bbde788749656b2a3da25c6d57/vault/token_store.go#L2372-L2375

When being used with globbing, this means the if allowed_policies contains say "foo-*" it would try and make a token with that literal policy, which of course would trigger a warning that the policy "foo-*" doesn't exist as per https://github.com/hashicorp/vault/blob/d6fea845bfb6b5bbde788749656b2a3da25c6d57/vault/token_store.go#L2665-L2673

Stripping out policies with globs in them at this point would prevent the warning with the behaviour being that when no policies are specified to the auth/token/create/:role endpoint that no glob-based policies are assigned (although any non-glob policies in the allow list would end up on the token).

It seemed cleaner to do it at that location, but feasibly the logic to strip out glob policies could be done at line 237 linked above where it sets the default policies to the role allowed policies.

Alternatively logic could be implemented for this default case to instead enumerate through the policies known to Vault and add any that match the globs present in the role - for example creating a token from a role with "foo-*" in allowed_policies and not specifying policies in the request would return a token with policies "foo-bar" and "foo-baz" if such policies exist in Vault. I don't have a use-case for this sort of behaviour so this wasn't included in the PR and I'm not sure how much of a desire there would be for this, but if need be I could switch the logic to do this instead.

Reply:
Thanks a lot for the super detailed explanation, I'll be honest I didn't at all pick up that allowed_policies gets populated into the token's policies, I thought it was a safeguard against globbed policies requested by the user ( and that, that might be a valid scenario we want to support )

For the question of where to put this, I think I would lean towards putting this around L2372-2375 with a comment noting why this change is required there. However, I think the latter approach you put forward, of enumerating out globbed policies in the allowlist, may actually be required to maintain feature parity with how this works today. I'm going to go ahead and chat with the team about how we want to proceed here, since this is not an easy choice, there are correctness implications to not enumerating these policies, and performance implications to looking up all possible policies on what is likely a frequently used code path. 

I will get back to you with an answer here soon as I can, and again, thanks so much for going over and explaining the problem!!

Reply:
Do note that if we went that way it'd make sense to only do lookup the policies if both no policies were listed in the request, and globbing is enabled. Nomad for example always lists a policy when creating a token from a role for each alloc, so it wouldn't trigger that code path. (This is actually my main use-case for this feature). So the impact might not be as bad as it initially seems.

I'm not too familiar with Vault's internals around policies, but if there was a more efficient way to list all policies that matched a set of globs rather than the simple approach of naive method of just enumerating through every policy and checking if they match that could make sense to do as well.

Reply:
Yeah, that is true, this would be a small edge case and unlikely to be used by people opting in to this functionality, however we would like to plan for that case too.
As far as performance, that after discussion I've found out is actually not a big deal, cause Vault keeps this cached in memory anyways, so it wouldn't require a costly iteration through storage.
However, after discussion I think we ended up being uncomfortable with either choice, be it ignoring the globbed policies when no policies are requested here, or adding all the policies matching the globs in the allowlist, cause that could accidentally give users more permissions than intended when setting up the policy.
Based on the fact that this approach ran into this problem, we'd like to actually try solving this problem in the other suggested approach. That would be, to have 2 new lists, `allowed_policies_glob` and `disallowed_policies_glob`, that would have globbed versions of the policies. What this achieves, is that we can still evaluate against globbed policies when a user requests some set of policies, with the new lists. The input would be checked against both glob and non-glob lists, for both the allow and disallow cases. At the same time, we can explicitly control the default policies we want added to a token created against the role, with the existing `allowed_policies` functionality.
I know this breaks from the PKI engine's pattern, however I think that should be fine cause the PKI engine's cert gen endpoints don't need this default population functionality.
How does this sound to you?

Reply:
That sounds reasonable. This would also remove the need for the `allow_glob_policies` flag as well since it's controlled by new lists instead.

I'll update the PR with that pattern. In the meantime I'll also update the stdlib PR to remove the RemoveGlobs func since that will no longer be needed either.

Reply:
Yep, we wouldn't need the `allow_glob_policies` flag with these 2 fields anymore. Kk, sounds good!

Reply:
Sweet, can you arrange for https://github.com/hashicorp/go-secure-stdlib/pull/8 to get merged. I've updated it to remove the now unneeded RemoveGlobs func. After that's in I can finish updating this PR with the new *_glob fields.

Reply:
Ah, I actually wanted to review this PR prior to merging that, could you maybe put a go.mod replace directive to use your secure-stdlib branch locally instead ? ( or alternately just paste the same code into files in this repo where it can be accessed ? ) I'd kinda prefer to review all the changes together before merging that part in. ( Just in case, say we run into any issues with this approach and end up not using that code, we don't need to revert )

Reply:
That makes sense, will do that shortly then